flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -197,13 +197,18 @@ BufferAndBacklog pollBuffer() {
 				return null;
 			}
 
+			boolean isExactlyOnceCheckpointBarrier = buffer.getDataType().isExactlyOnceCheckpointBarrier();
+			if (isExactlyOnceCheckpointBarrier) {
+				readView.blockOnCheckpointBarrier();
```

Comment:
The `readerView` maintains the blocked state, then I prefer to moving this judgement inside subpartition view. So we can remove this method `blockOnCheckpointBarrier` completely.

Reply:
To do so, we have three components to maintain the separate states to finalize the final available state.
`ResultSubpartition`:  data available via `BufferAndBacklog` internal state
`ResultSubpartitionView`: blocked state when polling barrier
`NetworkSequenceViewReader`: credit available

Reply:
As discussed, I moved the ```blockedByCheckpoint``` tag to subpartition.